### DISCUSSION TITLE
Tip: make a Windows command that has space character in path and takes argument work

### DISCUSSION THREAD

**Discussion by ghost** [gokcehan/lf#1276] [General]
Hi,

You can jump to TLDR to save time.

## background

I wanted to share my experience. I was trying to adding a new command for lf on Windows and encountered a problem. 

I went through issues, discussions, Wiki, documentation of this repository. I also Googled and checked Reddit, and couldn't find the answer. I was writing a new thread in this discussions forum to ask, when I accidentally find a way to make it work.

Since I think the problem I encountered was not documented somewhere on the net (or at least it's not easy to find), hence I wanted to share it here.

## the problem

When making a command for Windows platform, if the path of the executable doesn't contain a space character, or that no file is passed to the executable (as a command line argument), it's easier to make it work. For example, the following two commands work:

```
map D !C:\tools\neovim\nvim-win64\bin\nvim-qt.exe c:\work\todo.txt
```

```
map D !c:\Program Files\WinMerge\WinMergeU.exe
```

On the contrary, it takes special care to make it work. For example, the following doesn't work:

```
map D !c:\Program Files\WinMerge\WinMergeU.exe c:\work\todo.txt
```

I tried a lot of combinations, including the followings, and I only find one way to make it work, it's the one marked out:

```
#map D !c:\Program Files\WinMerge\WinMergeU.exe c:\work\todo.txt
#map D !'c:\Program Files\WinMerge\WinMergeU.exe c:\work\todo.txt'
#map D !'c:\Program Files\WinMerge\WinMergeU.exe' c:\work\todo.txt
#map D !'"c:\Program Files\WinMerge\WinMergeU.exe" c:\work\todo.txt'
#map D !"c:\Program Files\WinMerge\WinMergeU.exe" c:\work\todo.txt
#map D !"c:\Program Files\WinMerge\WinMergeU.exe c:\work\todo.txt"
#map D !'"c:\Program Files\WinMerge\WinMergeU.exe c:\work\todo.txt"'
map D !c:\"Program Files"\WinMerge\WinMergeU.exe c:\work\todo.txt   # works
#map D !'c:\"Program Files"\WinMerge\WinMergeU.exe' c:\work\todo.txt
#map D !'c:\"Program Files"\WinMerge\WinMergeU.exe c:\work\todo.txt'
```

Notice the point is not to quote the entire path, but a small part. In other words:

* "c:\\Program Files\\WinMerge\\WinMergeU.exe" -> NG
* c:\\"Program Files"\\WinMerge\\WinMergeU.exe -> OK

Maybe it's the supposed way, by design or maybe it's not the best way, I don't know.

If anyone has better way or more information, feel welcomed to share.

## TLDR

This is a way that I find working (quote only "Program Files", not entire path):

```
map D !c:\"Program Files"\WinMerge\WinMergeU.exe c:\work\todo.txt   # works
```

**Comment by gokcehan** [comment]
@kietfsek Thanks for coming up with a solution. I personally find quoting in Windows to be quite unintuitive. What is even worse is that some of the examples that does not work in lfrc (e.g. quoting the entire path) does seem to work in regular `cmd.exe`. I agree we should make this information more accessible. Feel free to add this information to somewhere appropriate in the wiki pages as well. Also, maybe we can consider adding an example to the example configuration files. Although, we should probably use a different example than `WinMergeU.exe` as it may not be available everywhere. Maybe an example with the standard Windows tools could be better, I'm not sure.

**Reply by ghost** [reply]
@gokcehan 

With the information @ilyagr provided, if the behavior is about to be changed, do you still think it's a good idea to add the information to Wiki?

**Reply by gokcehan** [reply]
@kietfsek Sorry for late reply as I was afk for a couple of weeks.

I have only briefly looked at the proposed patch but it is likely to be merged so I guess we should not add this information to the wiki as it will be invalidated soon.

**Comment by ilyagr** [comment]
We now have a proposed PR #1309 that is likely to change this behavior in future versions of lf, hopefully for the better. Feel free to test it out.

**Reply by joelim-work** [reply]
I don't have `WinMergeU.exe` installed, but I have tested the following (i.e. executables with a space in their path) against #1309, and it is working fine:

```
map x1 !"C:\Program Files\Go\bin\go.exe" version
map x2 !"C:\Program Files\Git\bin\git.exe" --version
```

**Reply by ghost** [reply]
I tested the PR with my test cases above.

Before the PR, only this works:

```text
map D !"c:\Program Files\WinMerge\WinMergeU.exe" c:\work\todo.txt
```

With the PR, these two work:

```text
map D !"c:\Program Files\WinMerge\WinMergeU.exe" c:\work\todo.txt
map D !c:\"Program Files"\WinMerge\WinMergeU.exe c:\work\todo.txt
```
